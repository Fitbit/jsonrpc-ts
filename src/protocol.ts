/**
 * JSON-RPC protocol types, along with functions for constructing and
 * parsing/validating JSON-RPC data structures.
 *
 * The constructor and parser functions all operate on native JavaScript
 * data structures. De/serializing from/to JSON must be handled
 * externally.
 */

import * as t from 'io-ts';

// Runtime types are variables which are used like types, which is
// reflected in their PascalCase naming scheme.
/* tslint:disable:variable-name */

/** The JSON primitive types (string, number, boolean, null) */
export const Primitive = t.union([t.string, t.number, t.boolean, t.null]);
export type Primitive = t.TypeOf<typeof Primitive>;

/** The JSON structured types (object, array) */
export const Structured = t.union([t.dictionary(t.string, t.any), t.Array]);
export type Structured = t.TypeOf<typeof Structured>;

/** Any of the JSON types; not undefined */
export const Some = t.union([Primitive, Structured]);
export type Some = t.TypeOf<typeof Some>;

/** JSON-RPC Request or Response id */
export const RPCID = t.union([t.Integer, t.string]);
export type RPCID = t.TypeOf<typeof RPCID>;

/** JSON-RPC Request params */
export const RPCParams = Structured;
export type RPCParams = t.TypeOf<typeof RPCParams>;

/** `error` member of a JSON-RPC Error object */
export const RPCError = t.intersection([
  t.interface({
    code: t.Integer,
    message: t.string,
  }),
  t.partial({
    data: Some,
  }),
]);
export type RPCError = t.TypeOf<typeof RPCError>;

/** JSON-RPC Request object */
export const RequestJSON = t.intersection([
  t.interface({
    jsonrpc: t.literal('2.0'),
    method: t.string,
    id: RPCID,
    params: t.union([RPCParams, t.undefined]),
  }),
  t.partial({
    result: t.undefined,
    error: t.undefined,
  }),
]);
export type RequestJSON = t.TypeOf<typeof RequestJSON>;

/** JSON-RPC Notification object */
export const NotificationJSON = t.intersection([
  t.interface({
    jsonrpc: t.literal('2.0'),
    method: t.string,
    params: t.union([RPCParams, t.undefined]),
  }),
  t.partial({
    id: t.undefined,
    result: t.undefined,
    error: t.undefined,
  }),
]);
export type NotificationJSON = t.TypeOf<typeof NotificationJSON>;

/** JSON-RPC Response object */
export const ResponseJSON = t.intersection([
  t.interface({
    jsonrpc: t.literal('2.0'),
    result: Some,
    id: t.union([t.Integer, t.string]),
  }),
  t.partial({
    method: t.undefined,
    params: t.undefined,
    error: t.undefined,
  }),
]);
export type ResponseJSON = t.TypeOf<typeof ResponseJSON>;

/** JSON-RPC Error object */
export const ErrorJSON = t.intersection([
  t.interface({
    jsonrpc: t.literal('2.0'),
    error: RPCError,
    id: t.union([RPCID, t.null]),
  }),
  t.partial({
    method: t.undefined,
    params: t.undefined,
    result: t.undefined,
  }),
]);
export type ErrorJSON = t.TypeOf<typeof ErrorJSON>;

/** Deserialized JSON-RPC Request */
export interface Request {
  kind: 'request';
  method: string;
  params?: Structured;
  id: RPCID;
}

/** Deserialized JSON-RPC Notification */
export interface Notification {
  kind: 'notification';
  method: string;
  params?: Structured;
}

/** Deserialized JSON-RPC Response */
export interface Response {
  kind: 'response';
  result: Some;
  id: RPCID;
}

/** Deserialized JSON-RPC Error */
export interface Error {
  kind: 'error';
  error: {
    code: number;
    message: string;
    data?: Some;
  };
  id: RPCID | null;
}

/** Parsed and categorized JSON-RPC message */
export type Message = Request | Notification | Response | Error;

/* tslint:enable:variable-name */

/** Parse an object as a JSON-RPC message. */
export function parse(obj: object): Message {
  if (RequestJSON.is(obj)) {
    return {
      kind: 'request',
      id: obj.id,
      method: obj.method,
      params: obj.params,
    };
  }

  if (NotificationJSON.is(obj)) {
    return {
      kind: 'notification',
      method: obj.method,
      params: obj.params,
    };
  }

  if (ResponseJSON.is(obj)) {
    return {
      kind: 'response',
      id: obj.id,
      result: obj.result,
    };
  }

  if (ErrorJSON.is(obj)) {
    return {
      kind: 'error',
      id: obj.id,
      error: obj.error,
    };
  }

  throw new TypeError('Not a valid JSON-RPC object');
}

/** Construct a JSON-RPC Request object. */
export function request(id: RPCID, method: string, params?: RPCParams): RequestJSON {
  if (!RPCID.is(id)) {
    throw new TypeError('Request ID must be a string or integer');
  }
  return { id, method, params, jsonrpc: '2.0' };
}

/** Construct a JSON-RPC Notification object. */
export function notification(method: string, params?: RPCParams): NotificationJSON {
  return { method, params, jsonrpc: '2.0' };
}

/** Construct a JSON-RPC Response. */
export function response(id: RPCID, result?: Some): ResponseJSON {
  if (!RPCID.is(id)) {
    throw new TypeError('Response ID must be a string or integer');
  }
    // tslint:disable-next-line:no-null-keyword
  const nulledResult = result !== undefined ? result : null;
  return { id, jsonrpc: '2.0', result: nulledResult };
}

/** Arguments for constructing a JSON-RPC Error object. */
export interface ErrorObject extends RPCError {
  id?: RPCID | null;
}

/** Construct a JSON-RPC Error object. */
export function error(error: ErrorObject): ErrorJSON {
  const { id, code, message, data } = error;
    // tslint:disable-next-line:no-null-keyword
  const nulledId = id !== undefined ? id : null;
    // tslint:disable-next-line:no-null-keyword
  if (nulledId !== null && !RPCID.is(id)) {
    throw new TypeError('Error ID must be string, integer, null or undefined');
  }
  if (!t.Integer.is(code)) {
    throw new TypeError('Error code must be an integer');
  }
  return { jsonrpc: '2.0', id: nulledId, error: { code, message, data } };
}
