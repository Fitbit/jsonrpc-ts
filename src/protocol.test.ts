// null is not interchangeable with undefined in JSON-RPC.
// tslint:disable:no-null-keyword

import * as jrpc from './protocol';

declare class Object {
  static entries(object: object): [string, any][];
}

describe('parse', () => {
  describe('a valid Request', () => {
    it.each([
      ['with a numeric id', { id: -7, method: 'foo.doBar/baz' }],
      ['with a string id', { id: 'three', method: 'hello' }],
      ['with params as empty array', { id: 55, method: 'foo', params: [] }],
      ['with params by-position', {
        id: 42,
        method: 'bar',
        params: [3, 'three', { three: 3 }, ['four', 'five']],
      }],
      ['with params by-name', {
        id: 8,
        method: 'baz',
        params: { foo: 'bar', quux: 'yes', 3: 5 } as object,
      }],
    ])('%s', (_, vector) => {
      const { id, method, params } = vector;
      // We can't reassemble the message from the spread object as any
      // properties that do not exist in the test vector would come into
      // existence on the reassembled message, with a value of
      // undefined. This would invalidate the tests as a property
      // that exists with the value undefined cannot be represented in
      // JSON.
      expect(jrpc.parse({ ...vector, jsonrpc: '2.0' }))
        .toEqual({ id, method, params, kind: 'request' });
    });
  });

  describe('a valid Notification', () => {
    it.each([
      ['with no params', { method: 'notifoo' }],
      ['with params as empty array', { method: 'blah', params: [] }],
      ['with params by-position', {
        method: 'a.method',
        params: [
          ['hello', 'bonjour', -7.2],
          'abc',
          { a: 3, b: 'four' },
        ],
      }],
      ['with params by-name', {
        method: 'yo',
        params: {
          a: 1,
          b: 'two',
          c: [3, 'four', 5],
          d: { e: 'f' },
        },
      }],
    ])('%s', (_, vector) => {
      const { method, params } = vector;
      expect(jrpc.parse({ ...vector, jsonrpc: '2.0' }))
        .toEqual({ method, params, kind: 'notification' });
    });
  });

  describe('a valid Response', () => {
    it.each([
      ['with numeric id', { id: -7, result: null }],
      ['with string id', { id: 'abc', result: null }],
      ['with numeric result', { id: 5, result: 3.7 }],
      ['with string result', { id: 48, result: 'hello' }],
      ['with boolean result', { id: 1, result: true }],
      ['with array result', {
        id: 86,
        result: ['one', 2, { three: 3 }, [4]],
      }],
      ['with object result', { id: 104, result: { yes: 'yup' } }],
    ])('%s', (_, vector) => {
      const { id, result } = vector;
      expect(jrpc.parse({ ...vector, jsonrpc: '2.0' }))
        .toEqual({ id, result, kind: 'response' });
    });
  });

  describe('a valid Error', () => {
    it.each([
      ['with no id or data', {
        id: null,
        error: { code: -37000, message: 'you dun goofed' },
      }],
      ['with numeric id, no data', {
        id: 7,
        error: { code: 42, message: 'everything' },
      }],
      ['with string id, no data', {
        id: 'asdf',
        error: { code: 0, message: 'zero' },
      }],
      ['with boolean data', {
        id: 2,
        error: { code: 33, message: 'm', data: false },
      }],
      ['with numeric data', {
        id: 3,
        error: { code: 34, message: 'm', data: 8 },
      }],
      ['with string data', {
        id: 4,
        error: { code: 35, message: 'q', data: 'nope' },
      }],
      ['with null data', {
        id: 88,
        error: { code: 123, message: 'yes', data: null },
      }],
      ['with array data', {
        id: 5,
        error: { code: 36, message: 'r', data: [1, 2, 'three'] },
      }],
      ['with object data', {
        id: 6,
        error: { code: 37, message: 's', data: { foo: 'bar' } },
      }],
    ])('%s', (_, vector) => {
      const { id, error } = vector;
      expect(jrpc.parse({ ...vector, jsonrpc: '2.0' }))
        .toEqual({ id, error, kind: 'error' });
    });
  });

  describe('rejects a malformed message', () => {
    const vectors: { [key: string]: any } = {
      'numeric value': 3,
      'boolean value': false,
      'string value': 'hello',
      'array of nonsense': [1, 2, 3],

      'request without jsonrpc key': {
        id: 7,
        method: 'hello',
        params: { foo: 'bar', baz: 5 },
      },
      'response without jsonrpc key': {
        id: 7,
        result: true,
      },
      'notification without jsonrpc key': {
        method: 'notify',
        params: [3, 4, 5],
      },
      'error without jsonrpc key': {
        id: 4,
        error: { code: 8, message: 'yo', data: 'yoyo' },
      },

      'request with wrong jsonrpc type': {
        jsonrpc: 2,
        id: 1344,
        method: 'argh',
      },
      'request with wrong jsonrpc version': {
        jsonrpc: '3.0',
        id: -2,
        method: 'the future is now',
      },
      'notification with wrong jsonrpc version': {
        jsonrpc: '1.4',
        method: 'parallel universe',
      },
      'response with wrong jsonrpc version': {
        jsonrpc: '5',
        id: 66,
      },
      'error with wrong jsonrpc version': {
        jsonrpc: 'two point oh',
        id: null,
        error: { code: 1, message: 'words' },
      },

      'request with fractional id': {
        jsonrpc: '2.0',
        id: 3.4,
        method: 'fractions!',
      },
      'request with structured id': {
        jsonrpc: '2.0',
        id: [3],
        method: 'nope',
      },
      'request with result key': {
        jsonrpc: '2.0',
        id: 8,
        method: 'hey',
        result: 'foo',
      },
      'request with params and result': {
        jsonrpc: '2.0',
        id: 115,
        method: 'uhoh',
        params: [1, 'two'],
        result: 'three',
      },
      'request with error key': {
        jsonrpc: '2.0',
        id: 3,
        method: 'woo',
        error: { code: 1, message: 'yeah' },
      },
      'request with numeric method': {
        jsonrpc: '2.0',
        id: 5,
        method: 6,
      },
      'request with boolean method': {
        jsonrpc: '2.0',
        id: 7,
        method: true,
      },
      'request with array method': {
        jsonrpc: '2.0',
        id: 6543,
        method: ['do this', 'then that'],
      },
      'request with object method': {
        jsonrpc: '2.0',
        id: 432,
        method: { do: 'that' },
      },

      'notification with result': {
        jsonrpc: '2.0',
        method: 'foo',
        result: 'wut',
      },
      'notification with params and result': {
        jsonrpc: '2.0',
        method: 'asdf',
        params: [1, 6],
        result: 'seven',
      },
      'notification with error key': {
        jsonrpc: '2.0',
        method: 'asdfasdf',
        error: { code: 33, message: 'yes' },
      },

      'response with error key': {
        jsonrpc: '2.0',
        id: 89,
        result: 'yes',
        error: { code: 1, message: 'waitaminute' },
      },
      'response with fractional id': {
        jsonrpc: '2.0',
        id: 6.4,
        result: 'foo',
      },

      'error with missing id': {
        jsonrpc: '2.0',
        error: { code: 2, message: 'hmm' },
      },
      'error with fractional id': {
        jsonrpc: '2.0',
        id: 8.1,
        error: { code: 6, message: 'yo' },
      },
      'error with missing code': {
        jsonrpc: '2.0',
        id: 7,
        error: { message: 'wut' },
      },
      'error with missing message': {
        jsonrpc: '2.0',
        id: 654,
        error: { code: 34, data: 'yup' },
      },
    };
    for (const [desc, vector] of Object.entries(vectors)) {
      test(desc, () => {
        expect(() => console.log(jrpc.parse(vector))).toThrow();
      });
    }
  });
});

describe('request serializer', () => {
  it('serializes a request with a numeric id', () =>
        expect(jrpc.request(-7, 'foo')).toMatchSnapshot());
  it('serializes a request with a string id', () =>
        expect(jrpc.request('foo', 'method')).toMatchSnapshot());
  it('serializes a request with an array of params', () =>
        expect(jrpc.request(-803, 'mmm', [3, 'seven'])).toMatchSnapshot());
  it('does not allow fractional IDs', () =>
        expect(() => jrpc.request(5.1, 'woop')).toThrow(TypeError));
});

describe('notification serializer', () => {
  it('serializes a notification with no params', () =>
        expect(jrpc.notification('aNotification')).toMatchSnapshot());
  it('serializes a notification with an array of params', () =>
        expect(jrpc.notification('asdffdsa', ['one', 2, 'three'])).toMatchSnapshot());
});

describe('response serializer', () => {
  it('serializes a response with a numeric id', () =>
        expect(jrpc.response(7654)).toMatchSnapshot());
  it('serializes a response with a string id', () =>
        expect(jrpc.response('ayedee', { result: { result: ['result'] } }))
            .toMatchSnapshot());
  it('does not allow fractional IDs', () =>
        expect(() => jrpc.response(1.1)).toThrow(TypeError));
});

describe('error serializer', () => {
  it('does not allow fractional IDs', () =>
        expect(() => jrpc.error({ id: 3.14, code: 1, message: '' }))
            .toThrow(TypeError));
  it('does not allow fractional codes', () =>
        expect(() => jrpc.error({ id: 3, code: 1.2, message: '' }))
            .toThrow(TypeError));
  it('serializes an error with a numeric id', () =>
        expect(jrpc.error({ id: 5, code: 1, message: '' }))
            .toMatchSnapshot());
  it('serializes an error with a null id', () =>
        expect(jrpc.error({ id: null, code: 0, message: '' }))
            .toMatchSnapshot());
  it('converts an undefined id to null', () =>
        expect(jrpc.error({ code: 1, message: '' }).id).toBeNull());
  it('serializes error data', () => {
    const data = { foo: 'bar', baz: ['a', 'b', 'c'] };
    expect(jrpc.error({ data, id: 3, code: 7, message: 'yarr' }))
      .toMatchObject({
        jsonrpc: '2.0',
        id: 3,
        error: { data, code: 7, message: 'yarr' },
      });
  });
});
