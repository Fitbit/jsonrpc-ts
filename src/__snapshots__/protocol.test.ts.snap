// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error serializer serializes an error with a null id 1`] = `
Object {
  "error": Object {
    "code": 0,
    "data": undefined,
    "message": "",
  },
  "id": null,
  "jsonrpc": "2.0",
}
`;

exports[`error serializer serializes an error with a numeric id 1`] = `
Object {
  "error": Object {
    "code": 1,
    "data": undefined,
    "message": "",
  },
  "id": 5,
  "jsonrpc": "2.0",
}
`;

exports[`notification serializer serializes a notification with an array of params 1`] = `
Object {
  "jsonrpc": "2.0",
  "method": "asdffdsa",
  "params": Array [
    "one",
    2,
    "three",
  ],
}
`;

exports[`notification serializer serializes a notification with no params 1`] = `
Object {
  "jsonrpc": "2.0",
  "method": "aNotification",
  "params": undefined,
}
`;

exports[`request serializer serializes a request with a numeric id 1`] = `
Object {
  "id": -7,
  "jsonrpc": "2.0",
  "method": "foo",
  "params": undefined,
}
`;

exports[`request serializer serializes a request with a string id 1`] = `
Object {
  "id": "foo",
  "jsonrpc": "2.0",
  "method": "method",
  "params": undefined,
}
`;

exports[`request serializer serializes a request with an array of params 1`] = `
Object {
  "id": -803,
  "jsonrpc": "2.0",
  "method": "mmm",
  "params": Array [
    3,
    "seven",
  ],
}
`;

exports[`response serializer serializes a response with a numeric id 1`] = `
Object {
  "id": 7654,
  "jsonrpc": "2.0",
  "result": null,
}
`;

exports[`response serializer serializes a response with a string id 1`] = `
Object {
  "id": "ayedee",
  "jsonrpc": "2.0",
  "result": Object {
    "result": Object {
      "result": Array [
        "result",
      ],
    },
  },
}
`;
